### PR TITLE
chore(schemas, topology): integrate event schema definitions

### PR DESCRIPTION
**Pull request by JeanMertz**
Follow-up to #11336.

This PR integrates schema definitions into the topology. After this PR is merged, every event has an extra `u16` attached to it, keeping track of the global schema to which the event adheres.

We don't do anything with this data yet, that's something that'll happen in a follow-up PR, when the `datadog_logs` sink needs to query the registry where it can find relevant fields to encode into the final payload.

The next step in this chain of PRs is to have the `remap` transform inform the VRL compiler of the field types within the event. This won't use the schema ID attached to events because it must do this at boot-time, and thus it traces back the chain of components feeding into the transform, asking them for information about the schema (and merging multiple schemas if two or more components feed into one descending component).

ref #11303

### CODE REVIEW

File: lib/vector-core/src/schema/registry.rs
Comment (fuchsnj) [inline lib/vector-core/src/schema/registry.rs:50]: Small nit to match language used elsewhere
```suggestion
    /// The 0th identifier is reserved for the default "empty" schema.
```

File: lib/vector-core/src/event/metadata.rs
Comment (fuchsnj) [inline lib/vector-core/src/event/metadata.rs:30]: What's the benefit of having this explicit redirection with the schema `Registry`, instead of using something like `Rc<schema::Definition>`? Is it just so the memory overhead per event is smaller?

Comment (fuchsnj) [reply lib/vector-core/src/event/metadata.rs:30]: Also how does this work for events stored to disk? Is this `Id` supposed to be valid for Vector restarts (with potentially different config?)

Comment (fuchsnj) [reply lib/vector-core/src/event/metadata.rs:30]: Why does the schema id need to be set on an individual event? Whatever is consuming this event should know what schema is expected of all events already.

**Review by fuchsnj — CHANGES_REQUESTED** [review]
I'd like more clarification on how the schema ID is being stored before I'm comfortable with this being merged

File: lib/vector-core/src/event/metadata.rs
Comment (jszwedko) [reply lib/vector-core/src/event/metadata.rs:30]: I share the questions around restart / reload and the trade-offs of registry lookup vs. `Rc`.

I believe we need to store the schema information on the event so that the sink is able to lookup where to find which semantic fields to encode in sink specific ways, but I'll let Jean comment further.

Comment (fuchsnj) [reply lib/vector-core/src/event/metadata.rs:30]: >I believe we need to store the schema information on the event so that the sink is able to lookup where to find which semantic fields to encode in sink specific ways
Isn't this something that is known at vector startup, purely based off of the config? All events going to a specific sink should always have the same schema (unless my understanding is wrong here)

Comment (fuchsnj) [reply lib/vector-core/src/event/metadata.rs:30]: I'm assuming the id is here on an event because there can be multiple sources for a single sink, and each source might have a different schema associated with it. However, I don't think these schema ids are stable enough for this to work. If vector is restarted with a different config, this could be pointing to a random schema. A restart isn't even required, since we support live-reloading of config.

Comment (JeanMertz) [reply lib/vector-core/src/event/metadata.rs:30]: > What's the benefit of having this explicit redirection with the schema `Registry`, instead of using something like `Rc<schema::Definition>`? Is it just so the memory overhead per event is smaller?

Correct.

> Also how does this work for events stored to disk? Is this `schema::Id` supposed to be valid for Vector restarts (with potentially different config?)

That is a good comment, and one that slipped my mind.

For starters, events stored on disk _will_ have the schema ID embedded in them because they are stored as metadata, which is attached to the event.

The next part is the schema registry. This is not (yet) stored on disk between restarts. I don't see a reason why we couldn't. We'd have to tie the schema back to the component (source or transform) it originated from, _and_ the `Output` within that component. Basically, we'd need to store `HashMap<OutputId, schema::Id>` and then separately store the entire `schema::Registry` to keep them in-sync.

The one complication there is if a component is renamed in the config, and/or if its configuration is changed, and that configuration influences the schema generated at boot (e.g. `datadog_agent`'s `multiple_outputs` config is one such example).

Having said all this, I _think_ the simplest course of action would be as follows:

1. We drop the `schema::Id` approach for now, instead storing the actual `schema::Definition`.
2. This will have a bit of an impact on performance, but we'll have to measure it to see if that's really visible in our soaks.
3. This allows us to serialize the schema definition without problems.
4. There is still the issue of sink schema requirements changing, in which case an in-flight event might get rejected once Vector starts up again. I don't see anything we can do there, other than having our dead-letter pipeline in the future, and logging an error about dropped events in the present. This should only happen if we change the schema requirements of a sink, which should only be done if the actual upstream service changed its requirements, which means existing events wouldn't be allowed anyway. Additionally, we can still mitigate this situation as much as possible by implementing backward-compatible changes, for example by continuing to accept old schemas, and automatically converting them to the new schema. As a final note on this, this is also one of the reasons sinks deal with "semantic meaning", not exact field requirements. It means that even if we need to serialize the data to a `msg` field whereas before it was named `message` by the upstream service, the sink will continue to look for a field with the semantic meaning of "message", so in both cases things will work as expected.

If performance turns out to be an issue, we can revisit this, and build a more robust registry + schema ID system that is stable across restarts/config changes.

Does this make sense to both of you @fuchsnj and @jszwedko?

> I'm assuming the id is here on an event because there can be multiple sources for a single sink, and each source might have a different schema associated with it.

That is correct. Checking if a sink can accept its inputs can be done at boot-time by traversing the individual inputs back to their source, but at runtime we still need to distinguish between the individual events flowing through the pipeline, and then fetching the semantic meaning for each individual event (two events might have the same semantic meaning attached to them, but both might have that meaning pointing at different fields, both events are valid, but have to be handled differently).

Comment (fuchsnj) [reply lib/vector-core/src/event/metadata.rs:30]: We had a discussion offline, but I'll put some of my comments here so it's more visible.

>For starters, events stored on disk will have the schema ID embedded in them because they are stored as metadata, which is attached to the event.

This is not true today. Event metadata is not currently serialized. (https://github.com/vectordotdev/vector/issues/7065).

For the stable ID issue, I believe the main problem today is when an event goes into a buffer (either in-memory or on-disk), the schema ID may not be valid once the event comes out. The "obvious" solution is to serialize the semantic-meaning with the event, but that has unnecessary overhead.

As a longer-term solution, I'd like to at least explore the idea of adding some kind of "translation" step after transforms run, but before the event is sent to a buffer. This transform would take both the event, and the semantic meaning, and output the exact information a specific sink needs to be able to process that event. This "translated output" would then be stable and no longer require the semantic meaning to be sent with it, and a sink would be able to process it without any additional information, even across restarts. The largest downside here being that we introduce a new type that's not an `Event`, which would likely required some work in the sinks to support this new type, and the buffers would need to understand how to store it. I think this is yet another reason we should be looking into more generic "Event" types.

**Review by fuchsnj — APPROVED** [review]
Approving now that we have https://github.com/vectordotdev/vector/issues/11416 to track my concerns. This work is not visible to users yet, so this is just unblocking the work built on top of this.

**Pull request merged**
